src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,425 @@
+#include "zx/FunctionalityConstruction.hpp"
+
+#include "Definitions.hpp"
+#include "ZXDiagram.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace zx {
+
+    bool FunctionalityConstruction::checkSwap(op_it it, op_it end, Qubit ctrl,
+                                              Qubit target, const qc::Permutation& p) {
+        if (it + 1 != end && it + 2 != end) {
+            auto& op1 = *(it + 1);
+            auto& op2 = *(it + 2);
+            if (op1->getType() == qc::OpType::X && op2->getType() == qc::OpType::X &&
+                op1->getNcontrols() == 1 && op2->getNcontrols() == 1) {
+                const auto tar1  = p.at(op1->getTargets().front());
+                const auto tar2  = p.at(op2->getTargets().front());
+                const auto ctrl1 = p.at((*op1->getControls().begin()).qubit);
+                const auto ctrl2 = p.at((*op2->getControls().begin()).qubit);
+                return ctrl == tar1 && tar1 == ctrl2 && target == ctrl1 && ctrl1 == tar2;
+            }
+        }
+        return false;
+    }
+
+    void FunctionalityConstruction::addZSpider(ZXDiagram& diag, zx::Qubit qubit,
+                                               std::vector<Vertex>& qubit_vertices,
```

Comment:
Would you mind fixing the snake case naming for the `qubit_vertices` parameter?
This holds for all occurrences in this file and the corresponding header.
I would assume `qubits` would be fine as well for naming the parameter since the `vertex` part is inherently incurred by the type itself.